### ISSUE TITLE
[Bug] Kimi-K2.6 Eagle3 reports zero accept rate

### ISSUE THREAD

**Issue by hvagadia** [sgl-project/sglang#27829]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

# Kimi-K2.6 Eagle3 reports zero accept rate

## Summary

`nvidia/Kimi-K2.6-NVFP4` + `nvidia/Kimi-K2.6-Eagle3` starts successfully in SGLang, loads the draft model, captures draft CUDA graphs, and serves requests, but every decode log line reports:

```text
accept len: 1.00, accept rate: 0.00
```

On the same B200 node, with the same SGLang image and same request prompts, `nvidia/Kimi-K2.5-NVFP4` + `nvidia/Kimi-K2.5-Thinking-Eagle3` shows healthy non-zero acceptance.

## Environment

- Hardware: 1x B200 node, 8 GPUs
- SGLang image: `lmsysorg/sglang:dev`
- SGLang git head: `bcd9c5a903ac0e0397433633c6218abc52e63d73`
- SGLang git log: `bcd9c5a903ac0e0397433633c6218abc52e63d73 2026-06-10 09:25:30 +0800 update pytorch-xpu to 2.12 (#27133)`

## Kimi-K2.6 command

```bash
python3 -m sglang.launch_server \
  --model-path nvidia/Kimi-K2.6-NVFP4 \
  --served-model-name nvidia/Kimi-K2.6-NVFP4 \
  --trust-remote-code \
  --tp 8 \
  --quantization modelopt_fp4 \
  --moe-runner-backend flashinfer_trtllm \
  --fp4-gemm-backend flashinfer_cutlass \
  --enable-flashinfer-allreduce-fusion \
  --disable-custom-all-reduce \
  --max-running-requests 256 \
  --chunked-prefill-size 32768 \
  --mem-fraction-static 0.85 \
  --speculative-algorithm EAGLE3 \
  --speculative-draft-model-path nvidia/Kimi-K2.6-Eagle3 \
  --speculative-draft-model-quantization unquant \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --host 0.0.0.0 \
  --port 8000
```

## Kimi-K2.5 control command

```bash
python3 -m sglang.launch_server \
  --model-path nvidia/Kimi-K2.5-NVFP4 \
  --served-model-name nvidia/Kimi-K2.5-NVFP4 \
  --trust-remote-code \
  --tp 8 \
  --quantization modelopt_fp4 \
  --moe-runner-backend flashinfer_trtllm \
  --fp4-gemm-backend flashinfer_cutlass \
  --enable-flashinfer-allreduce-fusion \
  --disable-custom-all-reduce \
  --max-running-requests 256 \
  --chunked-prefill-size 32768 \
  --mem-fraction-static 0.85 \
  --speculative-algorithm EAGLE3 \
  --speculative-draft-model-path nvidia/Kimi-K2.5-Thinking-Eagle3 \
  --speculative-draft-model-quantization unquant \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --host 0.0.0.0 \
  --port 8000
```

## Evidence

Kimi-K2.6 startup succeeds:

```text
Load weight end ... type=KimiK25ForConditionalGeneration ... quant_algo=NVFP4
Load weight end ... type=Eagle3DeepseekV2ForCausalLM
Capture draft cuda graph end
Capture draft extend cuda graph end
Application startup complete
```

Requests were sent with `curl` to `/v1/chat/completions` using the served model name. They were non-streaming single-user-message chat requests: four deterministic prompts with `temperature: 0.0`, `top_p: 1.0`, and `max_tokens` in `[768, 1536]`, plus one sampled prompt with `temperature: 0.7`, `top_p: 0.95`, and `max_tokens: 1024`. Prompt types included code generation, math reasoning, benchmark checklist generation, speculative decoding explanation, and a technical incident report. All returned HTTP 200.

Kimi-K2.6 telemetry:

```text
decode_rows=134
nonzero_accept_rows=0
mean_accept_rate=0.0
max_accept_rate=0.0
mean_accept_len=1.0
max_accept_len=1.0
```

Example Kimi-K2.6 decode lines:

```text
[2026-06-10 10:52:21 TP0] Decode batch, #running-req: 1, #token: 320, token usage: 0.00, accept len: 1.00, accept rate: 0.00, cuda graph: True, gen throughput (token/s): 141.41, #queue-req: 0
[2026-06-10 10:52:24 TP0] Decode batch, #running-req: 1, #token: 768, token usage: 0.00, accept len: 1.00, accept rate: 0.00, cuda graph: True, gen throughput (token/s): 141.24, #queue-req: 0
```

Control run on the same node with Kimi-K2.5:

```text
decode_rows=61
nonzero_accept_rows=61
mean_accept_rate=0.40
max_accept_rate=0.78
mean_accept_len=2.20
max_accept_len=3.35
```

Example Kimi-K2.5 decode line:

```text
[2026-06-10 10:13:48 TP0] Decode batch, #running-req: 1, #token: 832, token usage: 0.00, accept len: 2.50, accept rate: 0.50, cuda graph: True, gen throughput (token/s): 350.79, #queue-req: 0
```

**Comment by nvpohanh** [comment]
cc @thanhhao98

**Comment by thanhhao98** [comment]
I will check this

**Comment by thanhhao98** [comment]
Root cause: the nvidia/Kimi-K2.6-Eagle3 checkpoint didn't ship an lm_head weight — it's meant to be replicated from the target model. SGLang doesn't do that here, so the draft runs with an uninitialized lm_head → garbage draft logits → accept rate: 0.00.

I synced with the NVIDIA Model Optimizer team and they've updated the nvidia/Kimi-K2.6-Eagle3 checkpoint to include lm_head.weight. Pulling the latest revision resolves the zero-acceptance on current SGLang.

Even though the latest draft checkpoint already fixes this, I noticed that both vLLM and TRT-LLM borrow the target model's lm_head when the draft checkpoint doesn't ship one, so I'd suggest adding the same behavior to SGLang as defense-in-depth:


@kpham-sgl @nvpohanh what do you think?

**Comment by hvagadia** [comment]
Thanks for the fix, I can confirm I get non-zero AR now with updated model.

**Comment by kpham-sgl** [comment]
Thanks @thanhhao98  for the fix and thanks @hvagadia for confirming the fix. 

> Even though the latest draft checkpoint already fixes this, I noticed that both vLLM and TRT-LLM borrow the target model's lm_head when the draft checkpoint doesn't ship one, so I'd suggest adding the same behavior to SGLang as defense-in-depth:

I think draft ckpt should ship with `lm_head` explicitly so this is not needed
